File: backend/transaction-service/transaction-domain/transaction-application-service/src/main/java/com/saver/system/transaction/service/domain/TransactionCreateCommandHandler.java
package com.saver.system.transaction.service.domain;

import com.saver.system.transaction.service.domain.dto.create.CreateTransactionCommand;
import com.saver.system.transaction.service.domain.dto.create.CreateTransactionResponse;
import com.saver.system.transaction.service.domain.entity.Transaction;
import com.saver.system.transaction.service.domain.event.TransactionCreatedEvent;
import com.saver.system.transaction.service.domain.mapper.TransactionDataMapper;
import lombok.extern.slf4j.Slf4j;

@Slf4j
public class TransactionCreateCommandHandler {
    private final TransactionDataMapper transactionDataMapper;
    private final TransactionDomainServiceImpl transactionDomainService;

    TransactionCreateCommandHandler(TransactionDataMapper transactionDataMapper , TransactionDomainServiceImpl transactionDomainService) {
        this.transactionDataMapper = transactionDataMapper;
        this.transactionDomainService = transactionDomainService;
    }
    public CreateTransactionResponse createTransaction(CreateTransactionCommand createOrderCommand) {
        Transaction transaction = transactionDataMapper.createTransactionCommandToOrder(createOrderCommand);
        TransactionCreatedEvent orderCreatedEvent = transactionDomainService.validateAndInitiateTransaction(transaction);
        //  saveTransaction(transaction);

        log.info("Transaction created: {}", transaction);
        return transactionDataMapper.transactionToCreateTransactionResponse(orderCreatedEvent.getTransaction(),
                "transaction created successfully");
    }
//    private Transaction saveTransaction(Transaction transaction) {
//        Transaction transactionResult = transactionRepository.save(order);
//        if (transactionResult == null) {
//            log.error("Could not save order!");
//            throw new TransactionDomainException("Could not save order!");
//        }
//        log.info("Order is saved with id: {}", transactionResult.getId().getValue());
//        return transactionResult;
//    }
}

File: backend/Services/transactions/src/domain/domain-application-service/transaction-create-command-handler.ts
import { inject, injectable } from "inversify";
import { TYPES } from "../../container/types/inversify-types";
import { Transaction } from "../domain-core/entity/transaction";
import { TransactionCreatedEvent } from "../domain-core/event/transaction-created-event";
import { ITransactionDomainService } from "../domain-core/itransaction-domain-service";
import { CreateTransactionCommand } from "./dto/create/create-transaction-command";
import { CreateTransactionResponse } from "./dto/create/create-transaction-response";
import { TransactionDataMapper } from "./mapper/transaction-data-mapper";
import { ITransactionRepository } from "./ports/output/repository/itransaction-repository";


@injectable()
export class CreateTransactionCommandHandler {
  constructor(
    @inject(TYPES.TransactionDomainService)
    private transactionDomainService: ITransactionDomainService,
    @inject(TYPES.CashTransactionRepository)
    private cashTransactionRepository: ITransactionRepository
  ) {}

  public async createCashTransaction(
    createTransactionCommand: CreateTransactionCommand
  ): Promise<CreateTransactionResponse> {
    const transaction: Transaction =
      TransactionDataMapper.createCashTransactionCommandToTransaction(
        createTransactionCommand
      );
    const orderCreatedEvent: TransactionCreatedEvent =
      this.transactionDomainService.validateAndInitiateTransaction(transaction);
    await this.cashTransactionRepository.createCashTransaction(transaction);
    return TransactionDataMapper.cashTransactionToCreateTransactionResponse(
      orderCreatedEvent.getTransaction(),
      "Transaction created successfully"
    );
  }
}
